sympy/algebras/quaternion.py

```diff
@@ -0,0 +1,540 @@
+# https://en.wikipedia.org/wiki/Quaternion
+from __future__ import print_function
+
+from sympy.core.expr import Expr
+from sympy import Rational
+from sympy import re, im, conjugate
+from sympy import sqrt, sin, cos, acos, asin, exp, ln
+from sympy import trigsimp
+from sympy import diff, integrate
+from sympy import Matrix, Add, Mul
+from sympy import symbols, sympify
+from sympy.printing.latex import latex
+from sympy.printing import StrPrinter
+
+
+class Quaternion(Expr):
+    """Provides basic quaternion operations.
+    Quaternion objects can be instantiated as Quaternion(a, b, c, d)
+    as in (a + b*i + c*j + d*k).
+
+    Example
+    ========
+
+    >>> from sympy.algebras.quaternion import Quaternion
+    >>> q = Quaternion(1, 2, 3, 4)
+    >>> q
+    1 + 2*i + 3*j + 4*k
+
+    Quaternions over complex fields can be defined as :
+    ========
+    >>> from sympy.algebras.quaternion import Quaternion
+    >>> from sympy import symbols, I
+    >>> x = symbols('x')
+    >>> q1 = Quaternion(x, x**3, x, x**2, real_field = False)
+    >>> q2 = Quaternion(3 + 4*I, 2 + 5*I, 0, 7 + 8*I, real_field = False)
+    >>> q1
+    x + x**3*i + x*j + x**2*k
+    >>> q2
+    (3 + 4*I) + (2 + 5*I)*i + 0*j + (7 + 8*I)*k
+    """
+    _op_priority = 11.0
+
+    is_commutative = False
+
+    def __new__(cls, a=0, b=0, c=0, d=0, real_field = True):
+        a = sympify(a)
+        b = sympify(b)
+        c = sympify(c)
+        d = sympify(d)
+
+        if any(i.is_commutative is False for i in [a, b, c, d]):
+            raise ValueError("arguments have to be commutative")
+        else:
+            obj = Expr.__new__(cls, a, b, c, d)
+            obj._a = a
+            obj._b = b
+            obj._c = c
+            obj._d = d
+            obj._real_field = real_field
+            return obj
+
+    @property
+    def a(self):
+        return self._a
+
+    @property
+    def b(self):
+        return self._b
+
+    @property
+    def c(self):
+        return self._c
+
+    @property
+    def d(self):
+        return self._d
+    @property
+    def real_field(self):
+        return self._real_field
+
+    @classmethod
+    def from_axis_angle(cls, vector, angle):
+        """Returns a rotation quaternion given the axis and the angle of rotation.
+
+        Example
+        ========
+
+        >>> from sympy.algebras.quaternion import Quaternion
+        >>> from sympy import pi, sqrt
+        >>> q = Quaternion.from_axis_angle((sqrt(3)/3, sqrt(3)/3, sqrt(3)/3), 2*pi/3)
+        >>> q
+        1/2 + 1/2*i + 1/2*j + 1/2*k
+        """
+        (x, y, z) = vector
+        norm = sqrt(x**2 + y**2 + z**2)
+        (x, y, z) = (x / norm, y / norm, z / norm)
+        s = sin(angle * Rational(1, 2))
+        a = cos(angle * Rational(1, 2))
+        b = x * s
+        c = y * s
+        d = z * s
+
+        return cls(a, b, c, d).normalize()
+
+    @classmethod
+    def from_matrix(cls, M):
```

Comment:
this should be called `from_rotation_matrix`, i.e. there are many possible quaternion-matrix relationships.